pkg/services/screenshot/screenshot.go

```diff
@@ -0,0 +1,224 @@
+	ErrScreenshotsUnavailable = errors.New("screenshots unavailable")
+)
+
+// Screenshot represents a screenshot of a dashboard in Grafana.
+//
+// A screenshot can have an ImageOnDiskPath and an ImagePublicURL if the
+// screenshot is stored or disk and uploaded to a cloud storage service or
+// made accessible via the Grafana HTTP server.
+type Screenshot struct {
+	ImageOnDiskPath string
+	ImagePublicURL  string
+}
+
+// ScreenshotOptions are the options for taking a screenshot.
+// If the PanelID is zero-valued then a dashboard screenshot
+// is taken.
+type ScreenshotOptions struct {
+	DashboardUID     string
+	PanelID          int64
+	Width            int
+	Height           int
+	Theme            models.Theme
+	Timeout          time.Duration
+	UploadScreenshot bool
+}
+
+// SetDefaults sets default values for missing or invalid options.
+func (s ScreenshotOptions) SetDefaults() ScreenshotOptions {
+	if s.Width <= 0 {
+		s.Width = DefaultWidth
+	}
+	if s.Height <= 0 {
+		s.Height = DefaultHeight
+	}
+	switch s.Theme {
+	case models.ThemeDark, models.ThemeLight:
+	default:
+		s.Theme = DefaultTheme
+	}
+	if s.Timeout <= 0 {
+		s.Timeout = DefaultTimeout
+	}
+	return s
+}
+
+// ScreenshotService is an interface for taking screenshots.
+//go:generate mockgen -destination=mock.go -package=screenshot github.com/grafana/grafana/pkg/services/screenshot ScreenshotService
+type ScreenshotService interface {
+	Take(ctx context.Context, opts ScreenshotOptions) (*Screenshot, error)
+}
+
+// BrowserScreenshotService takes screenshots using a headless browser.
+type BrowserScreenshotService struct {
+	db sqlstore.Store
+	r  rendering.Service
+}
+
+func NewBrowserScreenshotService(db sqlstore.Store, r rendering.Service) ScreenshotService {
+	return &BrowserScreenshotService{
+		db: db,
+		r:  r,
+	}
+}
+
+func (s *BrowserScreenshotService) Take(ctx context.Context, opts ScreenshotOptions) (*Screenshot, error) {
+	q := models.GetDashboardQuery{Uid: opts.DashboardUID}
+	if err := s.db.GetDashboard(ctx, &q); err != nil {
+		return nil, err
+	}
+
+	opts = opts.SetDefaults()
+	renderOpts := rendering.Opts{
+		TimeoutOpts: rendering.TimeoutOpts{
+			Timeout: opts.Timeout,
+		},
+		AuthOpts: rendering.AuthOpts{
+			OrgID:   q.Result.OrgId,
+			OrgRole: models.ROLE_ADMIN,
+		},
+		Width:           opts.Width,
+		Height:          opts.Height,
+		Theme:           opts.Theme,
+		ConcurrentLimit: setting.AlertingRenderLimit,
+	}
+
+	if opts.PanelID == 0 {
+		// take a screenshot of the dashboard
+		renderOpts.Path = fmt.Sprintf("d/%s/%s/?orgId=%d", q.Result.Uid, q.Result.Slug, q.Result.OrgId)
+	} else {
+		renderOpts.Path = fmt.Sprintf("d-solo/%s/%s/?orgId=%d&panelId=%d", q.Result.Uid, q.Result.Slug, q.Result.OrgId, opts.PanelID)
+	}
+
+	result, err := s.r.Render(ctx, renderOpts, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to take screenshot: %w", err)
+	}
+
+	screenshot := Screenshot{
+		ImageOnDiskPath: result.FilePath,
+	}
+
+	if opts.UploadScreenshot {
```

Comment:
It will work the same as old alerting, which depends on the configuration in Hosted Grafana. It can either be Amazon S3, Google Cloud Storage bucket, Azure Blob storage, local storage, or disabled. I will update the design document!

Reply:
Thanks.